File: 07-week-6--foundational-data-structures/00-days-1-to-2--implement-a-linked-list/solutions/linked_list.js
class LinkedList {
  constructor(head = null) {
    this.head = head;
  }

  iterate(callback) {
    let count = 0;
    let temp = this.head;

    while (temp !== null) {
      const result = callback(temp, count);

      if (result === true) {
        return temp;
      }

      ++count;
      temp = temp.next;
    }

    return this.head;
  }

  // print each node's value on its own line
  // use your iterate method to be DRY! Don't get caught in the code rain, brrr.
  print() {
    this.iterate(node => console.log(node.value));
  }

  // find the node with the target value and return it
  // if not found return null, use your iterate method to be DRY!
  find(target) {
    let result = null;

    this.iterate(node => {
      if (node.value === target) {
        result = node;

        return true;
      }
    });

    return result;
  }

  // add the node to the start of the list, no nodes should be removed
  addFirst(node) {
    node.next = this.head;
    this.head = node;
  }

  // add node to end of list, no nodes should be removed
  // you may wish to use the iterate method
  addLast(node) {
    if (this.head === null) {
      this.head = node;
      return;
    }

    this.iterate(currNode => {
      if (currNode.next === null) {
        currNode.next = node;
        return true;
      }
    });
  }

  // remove the first Node in the list and update head
  // and return the removed node
  removeFirst() {
    const oldHead = this.head;

    if (this.head !== null) {
      this.head = this.head.next;
    }

    return oldHead;
  }

  // remove the tail node, iterate may be helpful
  // return the node you just removed
  removeLast() {
    if (this.head === null || this.head.next === null) {
      return this.removeFirst();
    }

    let oldTail = null;

    this.iterate(node => {
      if (node.next.next === null) {
        oldTail = node.next;
        node.next = null;
        return true;
      }
    });

    return oldTail;
  }

  // replace the node at the given index with the given node
  replace(idx, node) {
    if (idx === 0) {
      this.removeFirst();
      this.addFirst(node);
      return node;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        node.next = currNode.next.next;
        currNode.next = node;

        return true;
      }
    });

    return node;
  }

  // insert the node at the given index
  // no existing nodes should be removed or replaced
  insert(idx, node) {
    if (idx === 0) {
      this.addFirst(node);
      return;
    }

    this.iterate((currNode, count) => {
      if (count === idx - 1) {
        const oldNext = currNode.next;
        currNode.next = node;
        node.next = oldNext;

        return true;
      }
    });
  }

  // remove the node at the given index, and return it
  remove(idx) {
    if (idx === 0) {
      return this.removeFirst();
    }

    let oldNode = null;

    this.iterate((node, count) => {
      if (count === idx - 1) {
        oldNode = node.next;
        node.next = node.next.next;

        return true;
      }
    }); 

    return oldNode;
  }

  clear() {
    this.head = null;
  }
}

class Node {
  constructor(value = null, next = null) {
    this.value = value;
    this.next = next;
  }
}

if (require.main === module) {
  let head = new Node('one', new Node('two', new Node('three', new Node('four'))));
  let list = new LinkedList(head);
  let emptyList = new LinkedList();
  let oneItemList = new LinkedList(new Node('just one'));

  console.log("Print one to four");
  list.print();
  console.log("-----------------------------");

  console.log("Handle empty list print");
  emptyList.print();
  console.log("-----------------------------");

  console.log("Handle one item list print");
  oneItemList.print();
  console.log("-----------------------------");

  console.log(`Find four`);
  console.log(`${list.find('four').value}`);
  console.log(`Find non-existent value`);
  console.log(`Nothing: ${list.find(50)}`);
  console.log(`Find in empty list: ${emptyList.find(20)}`);
  console.log(`Find just one in one item list: ${oneItemList.find('just one').value}`);
  console.log(`Find nothing in one item list: ${oneItemList.find('nothing')}`);
  console.log("-----------------------------");

  console.log("Add zero as head");
  list.addFirst(new Node('zero'));
  list.print();
  console.log("-----------------------------");

  console.log("Add zero as head to empty list");
  emptyList.addFirst(new Node('zero'));
  emptyList.print();
  emptyList.head = null;
  console.log("-----------------------------");

  console.log("Add zero as head to one item list");
  oneItemList.addFirst(new Node('zero'));
  oneItemList.print();
  oneItemList.head = oneItemList.head.next;
  console.log("-----------------------------");

  console.log("Add five as tail");
  list.addLast(new Node('five'));
  list.print();
  console.log("-----------------------------");

  console.log("Add whaaa as tail to empty list");
  emptyList.addLast(new Node('whaaa'));
  emptyList.print();
  emptyList.head = null;
  console.log("-----------------------------");

  console.log("Add whaaa as tail to one item list");
  oneItemList.addLast(new Node('whaaa'));
  oneItemList.print();
  oneItemList.head.next = null;
  console.log("-----------------------------");

  console.log("Remove first node zero and return it");
  console.log(`${list.removeFirst().value} was removed`);
  list.print();
  console.log("-----------------------------");

  console.log("Remove first node from an empty list");
  console.log(`${emptyList.removeFirst()} was removed`);
  emptyList.print();
  console.log("-----------------------------");

  console.log("Remove first node from one item list");
  console.log(`${oneItemList.removeFirst().value} was removed`);
  oneItemList.print();
  oneItemList.head = new Node('just one');
  console.log("-----------------------------");

  console.log("Remove last node five and return it");
  console.log(`${list.removeLast().value} was removed`);
  list.print();
  console.log("-----------------------------");

  console.log("Remove last node from empty list and return it");
  console.log(`${emptyList.removeLast()} was removed`);
  emptyList.print();
  console.log("-----------------------------");

  console.log("Remove last node from one item list and return it");
  console.log(`${oneItemList.removeLast().value} was removed`);
  oneItemList.print();
  oneItemList.head = new Node('just one');
  console.log("-----------------------------");

  console.log("Replace node at index and return inserted node");
  console.log(`replace middle two with 2: ${list.replace(1, new Node('2')).value}`);
  list.print();
  console.log(`replace zeroth one with 1: ${list.replace(0, new Node('1')).value}`);
  list.print();
  console.log(`replace tail four with 4: ${list.replace(3, new Node('4')).value}`);
  list.print();
  console.log(`replace middle three with 3: ${list.replace(2, new Node('3')).value}`);
  list.print();
  console.log("-----------------------------");

  console.log("Insert node at index");
  console.log("Insert at 0");
  list.insert(0, new Node('zero'));
  list.print();
  list.removeFirst();

  console.log("Insert at 2");
  list.insert(2, new Node('two'));
  list.print();

  console.log("Insert at 4");
  list.insert(4, new Node('four'));
  list.print();

  console.log("Insert at 6");
  list.insert(6, new Node('six'));
  list.print();
  list.removeLast();

  console.log("Insert at 0 in empty list");
  emptyList.insert(0, new Node('zero'));
  emptyList.print();
  emptyList.removeFirst();
  console.log("-----------------------------");

  head = new Node('one', new Node('two', new Node('three', new Node('four'))));
  list = new LinkedList(head);

  console.log("Remove the node at the index and return it");
  console.log(`Remove two: ${list.remove(1).value}`);
  console.log(`Remove tail four: ${list.remove(2).value}`);
  console.log(`Remove three: ${list.remove(1).value}`);
  console.log(`Remove one: ${list.remove(0).value}`);
  list.print();
  console.log("-----------------------------");

  console.log("Clear a list");
  head = new Node('one', new Node('two', new Node('three', new Node('four'))));
  list = new LinkedList(head);
  list.clear();
  list.print();
}

module.exports = {
  Node, LinkedList
};

File: 07-week-6--foundational-data-structures/00-days-1-to-2--implement-a-linked-list/solutions/linked_list.rb
class LinkedList
  attr_accessor :head

  def initialize(head = nil)
    @head = head
  end

  def iterate
    count = 0
    temp = @head

    until temp.nil?
      yield(temp, count)      
      temp = temp.next_node
      count += 1
    end

    @head
  end

  def print
    iterate { |node| puts node.value }
  end

  def find(target)
    iterate do |node|
      return node if node.value == target
    end

    nil
  end

  def add_first(node)
    node.next_node = @head
    @head = node
  end

  def add_last(node)
    if @head.nil?
      @head = node
      return
    end

    iterate do |curr_node|
      if curr_node.next_node.nil?
        curr_node.next_node = node
        return
      end
    end
  end

  def remove_first
    old_head = @head
    @head = @head.next_node unless @head.nil?
    old_head
  end

  def remove_last
    return remove_first if @head.nil? || @head.next_node.nil?

    iterate do |node|
      if node.next_node.next_node.nil?
        old_tail = node.next_node
        node.next_node = nil
        return old_tail
      end
    end
  end

  def replace(idx, node)
    if idx.zero?
      remove_first
      add_first(node)
      return node
    end

    iterate do |curr_node, count|
      if count == idx - 1
        node.next_node = curr_node.next_node.next_node
        curr_node.next_node = node
        return node
      end
    end
  end

  def insert(idx, node)
    if idx.zero?
      add_first(node)
      return
    end

    iterate do |curr_node, count|
      if count == idx - 1
        old_next = curr_node.next_node
        curr_node.next_node = node
        node.next_node = old_next
        return
      end
    end
  end

  def remove(idx)
    if idx.zero?
      return remove_first
    end

    iterate do |node, count|
      if count == idx - 1
        old_node = node.next_node
        node.next_node = node.next_node.next_node
        return old_node
      end
    end
  end

  def clear
    @head = nil
  end
end

class Node
  attr_accessor :value, :next_node

  def initialize(value = nil, next_node = nil)
    @value = value
    @next_node = next_node
  end
end

if __FILE__ == $PROGRAM_NAME
  head = Node.new('one', Node.new('two', Node.new('three', Node.new('four'))))
  list = LinkedList.new(head)
  empty_list = LinkedList.new

  puts "Print one to four"
  list.print
  puts "-----------------------------"

  puts "Print empty list"
  empty_list.print
  puts "-----------------------------"

  puts "Find four"
  puts "#{list.find('four').value}"
  puts "Find non-existent value"
  puts "Nothing: #{list.find(50)}"
  puts "-----------------------------"

  puts "Find four in empty list"
  puts "#{empty_list.find('four')}"
  puts "-----------------------------"

  puts "Add zero as head"
  list.add_first(Node.new('zero'))
  list.print
  puts "-----------------------------"

  puts "Add zero as head to empty list"
  empty_list.add_first(Node.new('zero'))
  empty_list.print
  empty_list.head = nil
  puts "-----------------------------"

  puts "Add five as tail"
  list.add_last(Node.new('five'))
  list.print
  puts "-----------------------------"

  puts "Add five as tail to empty list"
  empty_list.add_last(Node.new('five'))
  empty_list.print
  empty_list.head = nil
  puts "-----------------------------"

  puts "Remove first node zero and return it"
  puts "#{list.remove_first.value} was removed"
  list.print
  puts "-----------------------------"

  puts "Remove first node from an empty list"
  puts "#{empty_list.remove_first} was removed"
  empty_list.print
  puts "-----------------------------"

  puts "Remove last node five and return it"
  puts "#{list.remove_last.value} was removed"
  list.print
  puts "-----------------------------"

  puts "Remove last node from empty list and return it"
  puts "#{empty_list.remove_last} was removed"
  empty_list.print
  puts "-----------------------------"

  puts "Replace node at index and return inserted node"
  puts "replace middle two with 2: #{list.replace(1, Node.new('2')).value}"
  list.print
  puts "replace zeroth one with 1: #{list.replace(0, Node.new('1')).value}"
  list.print
  puts "replace tail four with 4: #{list.replace(3, Node.new('4')).value}"
  list.print
  puts "replace middle three with 3: #{list.replace(2, Node.new('3')).value}"
  list.print
  puts "-----------------------------"

  puts "Insert node at index"
  puts "Insert at 0"
  list.insert(0, Node.new('zero'))
  list.print
  list.remove_first

  puts "Insert at 2"
  list.insert(2, Node.new('two'))
  list.print

  puts "Insert at 4"
  list.insert(4, Node.new('four'))
  list.print

  puts "Insert at 6"
  list.insert(6, Node.new('six'))
  list.print
  list.remove_last
  puts "-----------------------------"

  head = Node.new('one', Node.new('two', Node.new('three', Node.new('four'))))
  list = LinkedList.new(head)

  puts "Remove the node at the index and return it"
  puts "Remove two: #{list.remove(1).value}"
  puts "Remove tail four: #{list.remove(2).value}"
  puts "Remove three: #{list.remove(1).value}"
  puts "Remove one: #{list.remove(0).value}"
  list.print
  puts "-----------------------------"

  puts "Clear a list"
  head = Node.new('one', Node.new('two', Node.new('three', Node.new('four'))))
  list = LinkedList.new(head)
  list.clear
  list.print
end

# Please add your pseudocode to this file
# And a written explanation of your solution
